### PR TITLE
redis experiment

### PR DESCRIPTION
**Pull request by Geal**
implement caching in external storage for query plans, introspection and APQ. This is done as a multi level cache, first in memory with LRU then with a redis cluster backend. Since it is still experimental, it is opt-in through a Cargo feature

### CODE REVIEW

File: apollo-router/src/cache/storage.rs [hunk 1]
```diff
@@ -1,39 +1,273 @@
+use std::fmt;
 use std::hash::Hash;
 use std::sync::Arc;
 
 use lru::LruCache;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
 use tokio::sync::Mutex;
 
+pub(crate) trait KeyType:
```
Comment (BrynCooke) [inline apollo-router/src/cache/storage.rs:10]: Not seen this pattern before, nice :)

**Comment by BrynCooke** [comment]
In general this seems like a good first step, and as per the comments there'll be some changes as we add more storage options later.

File: apollo-router/src/cache/storage.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/cache/storage.rs:10]: I found that was cool too :)

File: apollo-router/src/cache/storage.rs [hunk 2]
```diff
@@ -234,12 +235,12 @@ impl RedisCacheStorage {
         println!("insert result {:?}", r);
     }
 
-    #[cfg(test)]
```
Comment (bnjjj) [inline apollo-router/src/cache/storage.rs]: If we don't need it we can remove it

Comment (Geal) [reply apollo-router/src/cache/storage.rs]: removed in [dacf151](https://github.com/apollographql/router/pull/2024/commits/dacf1514e317d49cd89785e384a021d83eef3ad5)

**Comment by Geal** [comment]
I think it's in a good shape now. Do we think things are missing to get it merged? What could be added after the merge (right now I'm thinking automatic reconnection to redis)

**Review by BrynCooke — APPROVED** [review]

**Pull request merged**